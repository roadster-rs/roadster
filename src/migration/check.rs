//! Utility methods to add check constraints to columns.

use sea_orm_migration::prelude::*;

/// Expression to check that a string column value is not empty.
pub fn str_not_empty<T>(name: T) -> SimpleExpr
where
    T: IntoIden + 'static,
{
    str_len_gt(name, 0)
}

/// Expression to check that a string column value's length is greater than the provided value.
pub fn str_len_gt<T>(name: T, len: u64) -> SimpleExpr
where
    T: IntoIden + 'static,
{
    Expr::expr(Func::char_length(Expr::col(name))).gt(len)
}

/// Expression to check that a string column value's length is greater than or equal to the
/// provided value.
pub fn str_len_gte<T>(name: T, len: u64) -> SimpleExpr
where
    T: IntoIden + 'static,
{
    Expr::expr(Func::char_length(Expr::col(name))).gte(len)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::assert_snapshot;
    use rstest::{fixture, rstest};
    use sea_orm_migration::schema::string;

    #[derive(DeriveIden)]
    pub(crate) enum Foo {
        Table,
        Bar,
    }

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn table_stmt() -> TableCreateStatement {
        Table::create().table(Foo::Table).to_owned()
    }

    #[rstest]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn str_not_empty(_case: TestCase, mut table_stmt: TableCreateStatement) {
        table_stmt.col(string(Foo::Bar).check(super::str_not_empty(Foo::Bar)));

        assert_snapshot!(table_stmt.to_string(PostgresQueryBuilder));
    }

    #[rstest]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn str_len_gt(_case: TestCase, mut table_stmt: TableCreateStatement) {
        table_stmt.col(string(Foo::Bar).check(super::str_len_gt(Foo::Bar, 1)));

        assert_snapshot!(table_stmt.to_string(PostgresQueryBuilder));
    }

    #[rstest]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn str_len_gte(_case: TestCase, mut table_stmt: TableCreateStatement) {
        table_stmt.col(string(Foo::Bar).check(super::str_len_gte(Foo::Bar, 1)));

        assert_snapshot!(table_stmt.to_string(PostgresQueryBuilder));
    }
}
