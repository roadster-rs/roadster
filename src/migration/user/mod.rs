use crate::migration::timestamp::m20240723_201404_add_update_timestamp_function;
use sea_orm_migration::prelude::*;

mod create_and_drop_table;
pub mod m20240714_203550_create_user_table_int_pk;
pub mod m20240714_203551_create_user_table_uuid_pk;
pub mod m20240723_070533_add_user_account_management_fields;
pub mod m20240724_005115_user_update_timestamp;
pub mod m20240729_000812_password_updated_at;
pub mod m20240729_002549_password_updated_at_function;
pub mod m20240729_002615_password_updated_at_trigger;
#[cfg(test)]
mod tests;

/// Contains the identifiers/fields created by all the `user` migrations.
#[derive(DeriveIden)]
pub(crate) enum User {
    Table,
    Id,
    Name,
    Username,
    Email,
    Password,
    /// When the user's password was updated. Defaults to the[crate::migration::timestamp::Timestamps::UpdatedAt]
    /// time. Useful in the event users' passwords may have been compromised and the application
    /// needs to enforce that users update their passwords.
    ///
    /// Updated automatically when the [User::Password] is updated, assuming the following
    /// migrations are applied:
    /// 1. [m20240729_002549_password_updated_at_function::Migration]
    /// 2. [m20240729_002615_password_updated_at_trigger::Migration]
    PasswordUpdatedAt,
    EmailConfirmationSentAt,
    EmailConfirmationToken,
    EmailConfirmedAt,
    LastSignInAt,
    RecoverySentAt,
    RecoveryToken,
    EmailChangeSentAt,
    /// Token sent to the new email to confirm it's a valid email and the user has access to it.
    EmailChangeTokenNew,
    /// Token sent to the current email to confirm the user authorized the email change.
    EmailChangeTokenCurrent,
    /// When the user was deleted.
    DeletedAt,
}

/// The collection of migrations defined to create a `user` table. Relevant [MigrationTrait]s
/// authored by `roadster` will be added here.
///
/// Note that the migration uses a `UUID` field for the `id` Primary Key field. If you would like
/// to use a `BIGINT` instead, you can do one of the following:
///
/// 1. Use the [m20240714_203550_create_user_table_int_pk::Migration] instead -- simply add it to
///    your main [MigratorTrait] implementation before the migrations from [UserMigrator].
/// 2. Add an `alter table` migration after the migrations from [UserMigrator].
#[non_exhaustive]
pub struct UserMigrator;

#[async_trait::async_trait]
impl MigratorTrait for UserMigrator {
    fn migrations() -> Vec<Box<dyn MigrationTrait>> {
        vec![
            Box::new(m20240714_203551_create_user_table_uuid_pk::Migration),
            Box::new(m20240723_070533_add_user_account_management_fields::Migration),
            Box::new(m20240723_201404_add_update_timestamp_function::Migration),
            Box::new(m20240724_005115_user_update_timestamp::Migration),
            Box::new(m20240729_000812_password_updated_at::Migration),
            Box::new(m20240729_002549_password_updated_at_function::Migration),
            Box::new(m20240729_002615_password_updated_at_trigger::Migration),
        ]
    }
}
