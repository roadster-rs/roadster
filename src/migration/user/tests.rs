use super::*;
use insta::assert_debug_snapshot;
use itertools::Itertools;

#[test]
fn user_migrator_migrations() {
    let user_migrations = UserMigrator::migrations()
        .into_iter()
        .map(|migration| migration.name().to_string())
        .collect_vec();
    assert_debug_snapshot!(user_migrations);
}

#[test]
fn user_migrator_migrations_no_int_pk() {
    let user_migrations = UserMigrator::migrations()
        .into_iter()
        .map(|migration| migration.name().to_string())
        .collect_vec();

    assert!(!user_migrations.contains(
        &m20240714_203550_create_user_table_int_pk::Migration::default()
            .name()
            .to_string()
    ))
}
