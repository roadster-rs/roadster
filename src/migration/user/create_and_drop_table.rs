use crate::migration::check::str_not_empty;
use crate::migration::schema::{pk_bigint_auto, pk_uuid, table};
use crate::migration::user::User;
use sea_orm_migration::{prelude::*, schema::*};

pub(crate) fn create_table_uuid_pk() -> TableCreateStatement {
    create_table(pk_uuid(User::Id))
}

pub(crate) fn create_table_int_pk() -> TableCreateStatement {
    create_table(pk_bigint_auto(User::Id))
}

pub(crate) fn create_table(pk_col: ColumnDef) -> TableCreateStatement {
    table(User::Table)
        .col(pk_col)
        .col(string(User::Name).check(str_not_empty(User::Name)))
        .col(string_uniq(User::Username).check(str_not_empty(User::Username)))
        .col(string_uniq(User::Email).check(str_not_empty(User::Email)))
        .col(string(User::Password))
        .to_owned()
}

pub(crate) fn drop_table() -> TableDropStatement {
    Table::drop().if_exists().table(User::Table).to_owned()
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use sea_orm::sea_query::PostgresQueryBuilder;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn create_table_uuid_pk() {
        let query = super::create_table_uuid_pk();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn create_table_int_pk() {
        let query = super::create_table_int_pk();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn drop_table() {
        let query = super::drop_table();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }
}
