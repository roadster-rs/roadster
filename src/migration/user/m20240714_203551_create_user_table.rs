use crate::migration::check::str_not_empty;
use crate::migration::schema::{pk_uuid, table};
use sea_orm_migration::prelude::*;
use sea_orm_migration::schema::{string, string_null, string_uniq, timestamp_with_time_zone_null};

#[derive(DeriveMigrationName)]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.create_table(create_table()).await
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.drop_table(drop_table()).await
    }
}

#[derive(DeriveIden)]
enum User {
    Table,
    Id,
    Name,
    Username,
    Email,
    Password,
    EmailConfirmationSentAt,
    EmailConfirmationToken,
    EmailConfirmedAt,
    LastSignInAt,
    RecoverySentAt,
    RecoveryToken,
    EmailChangeSentAt,
    /// Token sent to the new email to confirm it's a valid email and the user has access to it.
    EmailChangeTokenNew,
    /// Token sent to the current email to confirm the user authorized the email change.
    EmailChangeTokenCurrent,
    /// When the user was deleted.
    DeletedAt,
}

fn create_table() -> TableCreateStatement {
    table(User::Table)
        .col(pk_uuid(User::Id))
        .col(string(User::Name).check(str_not_empty(User::Name)))
        .col(string_uniq(User::Username).check(str_not_empty(User::Username)))
        .col(string_uniq(User::Email).check(str_not_empty(User::Email)))
        .col(string(User::Password))
        .col(timestamp_with_time_zone_null(User::EmailConfirmationSentAt))
        .col(string_null(User::EmailConfirmationToken))
        .col(timestamp_with_time_zone_null(User::EmailConfirmedAt))
        .col(timestamp_with_time_zone_null(User::LastSignInAt))
        .col(timestamp_with_time_zone_null(User::RecoverySentAt))
        .col(string_null(User::RecoveryToken))
        .col(timestamp_with_time_zone_null(User::EmailChangeSentAt))
        .col(string_null(User::EmailChangeTokenNew))
        .col(string_null(User::EmailChangeTokenCurrent))
        .col(timestamp_with_time_zone_null(User::DeletedAt))
        .to_owned()
}

fn drop_table() -> TableDropStatement {
    Table::drop().table(User::Table).to_owned()
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use sea_orm::sea_query::PostgresQueryBuilder;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn create_table() {
        let query = super::create_table();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn drop_table() {
        let query = super::drop_table();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }
}
