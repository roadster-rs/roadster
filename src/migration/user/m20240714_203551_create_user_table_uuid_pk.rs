//! Migrations to create a basic `user` table that contains the following fields:
//!
//! - Id (UUID)
//! - Name
//! - Username
//! - Email
//! - Password
//!
//! To add more fields, use the other migrations in the `user` mod.

use crate::migration::user::create_table::{create_table_uuid_pk, drop_table};
use sea_orm_migration::prelude::*;

#[derive(Default, DeriveMigrationName)]
#[non_exhaustive]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.create_table(create_table_uuid_pk()).await
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.drop_table(drop_table()).await
    }
}
