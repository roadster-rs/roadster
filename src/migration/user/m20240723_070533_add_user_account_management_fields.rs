//! Migrations to add some common account management fields to the `user` table. The following
//! fields are added:
//!
//! - EmailConfirmationSentAt
//! - EmailConfirmationToken
//! - EmailConfirmedAt
//! - LastSignInAt
//! - RecoverySentAt
//! - RecoveryToken
//! - EmailChangeSentAt
//! - EmailChangeTokenNew
//! - EmailChangeTokenCurrent
//! - DeletedAt

use crate::migration::user::User;
use sea_orm_migration::{prelude::*, schema::*};

#[derive(DeriveMigrationName)]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.alter_table(alter_table_add_columns()).await
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.alter_table(alter_table_drop_columns()).await
    }
}

fn alter_table_add_columns() -> TableAlterStatement {
    Table::alter()
        .table(User::Table)
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::EmailConfirmationSentAt))
        .add_column_if_not_exists(string_null(User::EmailConfirmationToken))
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::EmailConfirmedAt))
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::LastSignInAt))
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::RecoverySentAt))
        .add_column_if_not_exists(string_null(User::RecoveryToken))
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::EmailChangeSentAt))
        .add_column_if_not_exists(string_null(User::EmailChangeTokenNew))
        .add_column_if_not_exists(string_null(User::EmailChangeTokenCurrent))
        .add_column_if_not_exists(timestamp_with_time_zone_null(User::DeletedAt))
        .to_owned()
}

fn alter_table_drop_columns() -> TableAlterStatement {
    Table::alter()
        .table(User::Table)
        .drop_column(User::EmailConfirmationSentAt)
        .drop_column(User::EmailConfirmationToken)
        .drop_column(User::EmailConfirmedAt)
        .drop_column(User::LastSignInAt)
        .drop_column(User::RecoverySentAt)
        .drop_column(User::RecoveryToken)
        .drop_column(User::EmailChangeSentAt)
        .drop_column(User::EmailChangeTokenNew)
        .drop_column(User::EmailChangeTokenCurrent)
        .drop_column(User::DeletedAt)
        .to_owned()
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use sea_orm::sea_query::PostgresQueryBuilder;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn alter_table_add_columns() {
        let query = super::alter_table_add_columns();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn alter_table_drop_columns() {
        let query = super::alter_table_drop_columns();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }
}
