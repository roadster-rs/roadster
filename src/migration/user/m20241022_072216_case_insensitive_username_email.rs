use crate::migration::collation::Collation;
use crate::migration::user::User;
use sea_orm::{DbBackend, Statement, TransactionTrait};
use sea_orm_migration::prelude::*;

#[derive(DeriveMigrationName)]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        let db = manager.get_connection();
        let transaction = db.begin().await?;

        for statement in up_statements(manager.get_database_backend()) {
            transaction.execute(statement).await?;
        }

        transaction.commit().await?;

        Ok(())
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        let db = manager.get_connection();
        let transaction = db.begin().await?;

        for statement in down_statements(manager.get_database_backend()) {
            transaction.execute(statement).await?;
        }

        transaction.commit().await?;

        Ok(())
    }
}

fn up_statements(db_backend: DbBackend) -> Vec<Statement> {
    vec![
        alter_table_set_case_insensitive_collation(db_backend, User::Username),
        alter_table_set_case_insensitive_collation(db_backend, User::Email),
    ]
}

fn alter_table_set_case_insensitive_collation(db_backend: DbBackend, column: User) -> Statement {
    Statement::from_string(
        db_backend,
        format!(
            r#"ALTER table "{}" ALTER COLUMN "{}" type {} COLLATE {}"#,
            User::Table.to_string(),
            column.to_string(),
            "varchar",
            Collation::CaseInsensitive.to_string(),
        ),
    )
}

fn down_statements(db_backend: DbBackend) -> Vec<Statement> {
    vec![
        alter_table_reset_collation(db_backend, User::Username),
        alter_table_reset_collation(db_backend, User::Email),
    ]
}

fn alter_table_reset_collation(db_backend: DbBackend, column: User) -> Statement {
    Statement::from_string(
        db_backend,
        format!(
            r#"ALTER table "{}" ALTER COLUMN "{}" type {} COLLATE "{}""#,
            User::Table.to_string(),
            column.to_string(),
            "varchar",
            Collation::Default.to_string(),
        ),
    )
}

#[cfg(test)]
mod tests {
    use crate::testing::snapshot::TestCase;
    use insta::assert_debug_snapshot;
    use rstest::{fixture, rstest};
    use sea_orm::DbBackend;

    #[fixture]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(DbBackend::Postgres)]
    #[case(DbBackend::MySql)]
    #[case(DbBackend::Sqlite)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn up_statements(_case: TestCase, #[case] db_backend: DbBackend) {
        let query = super::up_statements(db_backend);

        assert_debug_snapshot!(query);
    }

    #[rstest]
    #[case(DbBackend::Postgres)]
    #[case(DbBackend::MySql)]
    #[case(DbBackend::Sqlite)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn down_statements(_case: TestCase, #[case] db_backend: DbBackend) {
        let query = super::down_statements(db_backend);

        assert_debug_snapshot!(query);
    }
}
