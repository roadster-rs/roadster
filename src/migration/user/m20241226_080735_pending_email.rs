use crate::migration::user::User;
use sea_orm_migration::{prelude::*, schema::*};

#[derive(DeriveMigrationName)]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.alter_table(alter_table_add_columns()).await
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager.alter_table(alter_table_drop_columns()).await
    }
}

fn alter_table_add_columns() -> TableAlterStatement {
    Table::alter()
        .table(User::Table)
        .add_column_if_not_exists(string_null(User::PendingEmail))
        .to_owned()
}

fn alter_table_drop_columns() -> TableAlterStatement {
    Table::alter()
        .table(User::Table)
        .drop_column(User::PendingEmail)
        .to_owned()
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use sea_orm::sea_query::PostgresQueryBuilder;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn alter_table_add_columns() {
        let query = super::alter_table_add_columns();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn alter_table_drop_columns() {
        let query = super::alter_table_drop_columns();

        assert_snapshot!(query.to_string(PostgresQueryBuilder));
    }
}
