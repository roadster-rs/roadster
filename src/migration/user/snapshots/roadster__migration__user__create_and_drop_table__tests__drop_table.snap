---
source: src/migration/user/create_and_drop_table.rs
expression: query.to_string(PostgresQueryBuilder)
---
DROP TABLE IF EXISTS "user"
